File: app/licensing.py
"""Regras de licenciamento — cada software possui licença independente."""
from __future__ import annotations

import re
import secrets
import string
from datetime import datetime, timedelta, timezone
from typing import Any

LICENSE_KEY_LEN = 25
LICENSE_ALPHABET = string.ascii_uppercase + string.digits
LICENSE_KEY_PATTERN = re.compile(rf"^[A-Z0-9]{{{LICENSE_KEY_LEN}}}$")

PRODUCT_CLOUD = "cloud"
PRODUCT_LAB = "lab"
PRODUCT_VDE = "vde"
ALLOWED_PRODUCTS = {PRODUCT_CLOUD, PRODUCT_LAB, PRODUCT_VDE}

PERIOD_TRIAL = "trial"
PERIOD_1Y = "1y"
PERIOD_2Y = "2y"
PERIOD_3Y = "3y"
PERIOD_4Y = "4y"
PERIOD_5Y = "5y"
ALLOWED_PERIODS = {PERIOD_TRIAL, PERIOD_1Y, PERIOD_2Y, PERIOD_3Y, PERIOD_4Y, PERIOD_5Y}

PERIOD_DAYS: dict[str, int] = {
    PERIOD_TRIAL: 30,
    PERIOD_1Y: 365,
    PERIOD_2Y: 730,
    PERIOD_3Y: 1095,
    PERIOD_4Y: 1460,
    PERIOD_5Y: 1825,
}

PRODUCT_LABELS = {
    PRODUCT_CLOUD: "Produto 01",
    PRODUCT_LAB: "Produto 02",
    PRODUCT_VDE: "Produto 03",
}

PERIOD_LABELS = {
    PERIOD_TRIAL: "Teste (30 dias)",
    PERIOD_1Y: "1 ano",
    PERIOD_2Y: "2 anos",
    PERIOD_3Y: "3 anos",
    PERIOD_4Y: "4 anos",
    PERIOD_5Y: "5 anos",
}

PAYMENT_PLAN_MONTHLY = "monthly"
PAYMENT_PLAN_SEMIANNUAL = "semiannual"
PAYMENT_PLAN_ANNUAL = "annual"
ALLOWED_PAYMENT_PLANS = {PAYMENT_PLAN_MONTHLY, PAYMENT_PLAN_SEMIANNUAL, PAYMENT_PLAN_ANNUAL}
PAYMENT_PLAN_LABELS = {
    PAYMENT_PLAN_MONTHLY: "Mensal",
    PAYMENT_PLAN_SEMIANNUAL: "Semestral",
    PAYMENT_PLAN_ANNUAL: "Anual",
}

ALERT_MILESTONES = (20, 15, 7, 3, 2, 1)

STATUS_PENDING = "pending"
STATUS_ACTIVE = "active"
STATUS_GRACE = "grace"
STATUS_BLOCKED = "blocked"
STATUS_CANCELLED = "cancelled"
STATUS_REVOKED = "revoked"
STATUS_EXPIRED = "expired"

API_PRODUCT_ALIASES = {
    "cloud": PRODUCT_CLOUD,
    "erp": PRODUCT_CLOUD,
    "lab": PRODUCT_LAB,
    "vde": PRODUCT_VDE,
    "vde_incorporadora": PRODUCT_VDE,
}


def now_utc() -> datetime:
    return datetime.now(timezone.utc)


def now_ts() -> str:
    return now_utc().strftime("%Y-%m-%d %H:%M:%S")


def parse_ts(value: str | datetime | None) -> datetime | None:
    if value is None:
        return None
    if isinstance(value, datetime):
        return value if value.tzinfo else value.replace(tzinfo=timezone.utc)
    raw = str(value).strip().replace("T", " ").replace("Z", "")
    for fmt in ("%Y-%m-%d %H:%M:%S", "%Y-%m-%d"):
        try:
            dt = datetime.strptime(raw[:19] if " " in fmt else raw[:10], fmt)
            return dt.replace(tzinfo=timezone.utc)
        except ValueError:
            continue
    try:
        return datetime.fromisoformat(raw).replace(tzinfo=timezone.utc)
    except ValueError:
        return None


def generate_license_key() -> str:
    return "".join(secrets.choice(LICENSE_ALPHABET) for _ in range(LICENSE_KEY_LEN))


def generate_lab_secret() -> str:
    return secrets.token_hex(32)


def normalize_license_key(key: str | None) -> str:
    return re.sub(r"[^A-Za-z0-9]", "", key or "").upper()[:LICENSE_KEY_LEN]


def is_valid_license_key_format(key: str) -> bool:
    normalized = normalize_license_key(key)
    return bool(LICENSE_KEY_PATTERN.match(normalized))


def compute_ends_at(start: datetime, periodo: str) -> datetime:
    days = PERIOD_DAYS.get(periodo, PERIOD_DAYS[PERIOD_TRIAL])
    return start + timedelta(days=days)


def normalize_api_product(product: str) -> str:
    return API_PRODUCT_ALIASES.get((product or "").strip().lower(), (product or "").strip().lower())


def product_includes_erp(produto: str) -> bool:
    return produto == PRODUCT_CLOUD


def product_includes_lab(produto: str) -> bool:
    return produto == PRODUCT_LAB


def product_includes_vde(produto: str) -> bool:
    return produto == PRODUCT_VDE


def product_matches_license(api_product: str, license_produto: str) -> bool:
    """Licença exclusiva por software — sem compartilhamento entre produtos."""
    normalized = normalize_api_product(api_product)
    legacy = license_produto.strip().lower()
    if legacy == "cloud_lab":
        return False
    return normalized == legacy


def days_until(target: datetime | None, ref: datetime) -> int:
    if not target:
        return 0
    return max(0, (target.date() - ref.date()).days)


def compute_license_validity(
    *,
    manual_status: str,
    ends_at: str | datetime | None,
    now: datetime | None = None,
) -> dict[str, Any]:
    ref = now or now_utc()
    if manual_status in {STATUS_CANCELLED, STATUS_REVOKED}:
        return {
            "licenseStatus": manual_status,
            "licenseExpired": True,
            "daysRemaining": 0,
            "validForSoftware": False,
        }

    due = parse_ts(ends_at)
    if not due:
        return {
            "licenseStatus": STATUS_PENDING,
            "licenseExpired": False,
            "daysRemaining": 0,
            "validForSoftware": False,
        }

    remaining = days_until(due, ref)
    expired = ref > due
    return {
        "licenseStatus": STATUS_EXPIRED if expired else STATUS_ACTIVE,
        "licenseExpired": expired,
        "daysRemaining": 0 if expired else remaining,
        "validForSoftware": not expired,
    }


def compute_payment_phase(
    *,
    manual_status: str,
    payment_due_at: str | datetime | None,
    block_after_days: int,
    cancel_after_days: int,
    now: datetime | None = None,
) -> dict[str, Any]:
    ref = now or now_utc()
    if manual_status in {STATUS_CANCELLED, STATUS_REVOKED}:
        return {
            "paymentPhase": manual_status,
            "paymentStatus": manual_status,
            "daysOverdue": 0,
            "block_at": None,
            "cancel_eligible_at": None,
            "message": "Cliente cancelado ou licença revogada.",
        }

    due = parse_ts(payment_due_at)
    if not due:
        return {
            "paymentPhase": STATUS_PENDING,
            "paymentStatus": STATUS_PENDING,
            "daysOverdue": 0,
            "block_at": None,
            "cancel_eligible_at": None,
            "message": "Aguardando definição de pagamento.",
        }

    block_at = due + timedelta(days=block_after_days)
    cancel_at = due + timedelta(days=cancel_after_days)
    days_overdue = max(0, (ref - due).days)

    if ref <= due:
        return {
            "paymentPhase": STATUS_ACTIVE,
            "paymentStatus": STATUS_ACTIVE,
            "daysOverdue": 0,
            "block_at": block_at.isoformat(),
            "cancel_eligible_at": cancel_at.isoformat(),
            "message": "Pagamento em dia.",
        }

    if ref <= block_at:
        grace_left = block_after_days - days_overdue
        return {
            "paymentPhase": STATUS_GRACE,
            "paymentStatus": STATUS_GRACE,
            "daysOverdue": days_overdue,
            "block_at": block_at.isoformat(),
            "cancel_eligible_at": cancel_at.isoformat(),
            "message": (
                f"Pagamento vencido há {days_overdue} dia(s). "
                f"Carência comercial: {grace_left} dia(s) restante(s)."
            ),
        }

    if ref <= cancel_at:
        return {
            "paymentPhase": STATUS_BLOCKED,
            "paymentStatus": STATUS_BLOCKED,
            "daysOverdue": days_overdue,
            "block_at": block_at.isoformat(),
            "cancel_eligible_at": cancel_at.isoformat(),
            "message": (
                f"Bloqueio comercial por inadimplência ({days_overdue} dias após vencimento do pagamento)."
            ),
        }

    return {
        "paymentPhase": "cancel_eligible",
        "paymentStatus": STATUS_BLOCKED,
        "daysOverdue": days_overdue,
        "block_at": block_at.isoformat(),
        "cancel_eligible_at": cancel_at.isoformat(),
        "message": (
            f"Elegível para cancelamento comercial ({days_overdue} dias). "
            f"Limite: {cancel_after_days} dias após vencimento do pagamento."
        ),
    }


def compute_alert_level(
    *,
    license_expired: bool,
    days_remaining: int,
    payment_phase: str,
) -> str:
    if license_expired:
        return "expired"
    if payment_phase in {STATUS_BLOCKED, "cancel_eligible"}:
        return "critical"
    if payment_phase == STATUS_GRACE:
        return "warning"
    if days_remaining in ALERT_MILESTONES:
        return "warning"
    if days_remaining <= 3:
        return "critical"
    return "none"


def compute_effective_status(
    *,
    manual_status: str,
    ends_at: str | datetime | None,
    payment_due_at: str | datetime | None = None,
    block_after_days: int = 30,
    cancel_after_days: int = 45,
    now: datetime | None = None,
) -> dict[str, Any]:
    ref = now or now_utc()
    validity = compute_license_validity(manual_status=manual_status, ends_at=ends_at, now=ref)
    payment = compute_payment_phase(
        manual_status=manual_status,
        payment_due_at=payment_due_at or ends_at,
        block_after_days=block_after_days,
        cancel_after_days=cancel_after_days,
        now=ref,
    )

    valid_for_software = validity["validForSoftware"] and manual_status not in {
        STATUS_REVOKED,
        STATUS_CANCELLED,
    }
    if payment["paymentPhase"] in {STATUS_BLOCKED, "cancel_eligible"}:
        valid_for_software = False

    alert_level = compute_alert_level(
        license_expired=validity["licenseExpired"],
        days_remaining=validity["daysRemaining"],
        payment_phase=payment["paymentPhase"],
    )

    status = payment["paymentPhase"]
    if validity["licenseExpired"]:
        status = STATUS_EXPIRED

    return {
        **validity,
        **payment,
        "status": status,
        "phase": status,
        "daysRemaining": validity["daysRemaining"],
        "daysOverdue": payment["daysOverdue"],
        "validForSoftware": valid_for_software,
        "alertLevel": alert_level,
        "licenseExpired": validity["licenseExpired"],
        "paymentPhase": payment["paymentPhase"],
    }


def sync_status_for_product_db(effective: dict[str, Any]) -> str:
    phase = effective.get("phase") or effective.get("status")
    if phase in {STATUS_CANCELLED, STATUS_REVOKED}:
        return STATUS_REVOKED
    if effective.get("licenseExpired"):
        return STATUS_BLOCKED
    if phase in {STATUS_BLOCKED, "cancel_eligible"}:
        return STATUS_BLOCKED
    if phase == STATUS_GRACE:
        return STATUS_ACTIVE
    if phase == STATUS_ACTIVE and effective.get("validForSoftware"):
        return STATUS_ACTIVE
    return STATUS_PENDING


def format_remaining_counter(days: int) -> str:
    if days <= 0:
        return "Vencido"
    years, rem = divmod(days, 365)
    months, d = divmod(rem, 30)
    parts = []
    if years:
        parts.append(f"{years} ano{'s' if years > 1 else ''}")
    if months:
        parts.append(f"{months} mês{'es' if months > 1 else ''}")
    if d or not parts:
        parts.append(f"{d} dia{'s' if d != 1 else ''}")
    return ", ".join(parts)

File: apps/web/src/domain/licensing.ts
/**
 * Regras de licenciamento — portadas do gerador Excellence Dental Cloud.
 */
import { randomBytes } from "crypto";

export const LICENSE_KEY_LEN = 25;
export const LICENSE_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

export const PRODUCT_CLOUD = "cloud";
export const PRODUCT_LAB = "lab";
export const PRODUCT_VDE = "vde";
export const PRODUCT_LIMPEZA = "limpeza";
export const PRODUCT_OUTROS = "outros";

export const PERIOD_TRIAL = "trial";
export const PERIOD_1Y = "1y";
export const PERIOD_2Y = "2y";
export const PERIOD_3Y = "3y";
export const PERIOD_4Y = "4y";
export const PERIOD_5Y = "5y";

export const PAYMENT_PLAN_MONTHLY = "monthly";
export const PAYMENT_PLAN_SEMIANNUAL = "semiannual";
export const PAYMENT_PLAN_ANNUAL = "annual";

export const ALLOWED_PERIODS = new Set([
  PERIOD_TRIAL,
  PERIOD_1Y,
  PERIOD_2Y,
  PERIOD_3Y,
  PERIOD_4Y,
  PERIOD_5Y,
]);

export const ALLOWED_PAYMENT_PLANS = new Set([
  PAYMENT_PLAN_MONTHLY,
  PAYMENT_PLAN_SEMIANNUAL,
  PAYMENT_PLAN_ANNUAL,
]);

export const PERIOD_DAYS: Record<string, number> = {
  [PERIOD_TRIAL]: 30,
  [PERIOD_1Y]: 365,
  [PERIOD_2Y]: 730,
  [PERIOD_3Y]: 1095,
  [PERIOD_4Y]: 1460,
  [PERIOD_5Y]: 1825,
};

export const PRODUCT_LABELS: Record<string, string> = {
  [PRODUCT_CLOUD]: "Excellence Dental Cloud",
  [PRODUCT_LAB]: "Dental Lab",
  [PRODUCT_VDE]: "VDE Incorporadora",
  [PRODUCT_LIMPEZA]: "Script de Limpeza completo",
  [PRODUCT_OUTROS]: "Outros sistemas",
};

export const PERIOD_LABELS: Record<string, string> = {
  [PERIOD_TRIAL]: "Teste (30 dias)",
  [PERIOD_1Y]: "1 ano",
  [PERIOD_2Y]: "2 anos",
  [PERIOD_3Y]: "3 anos",
  [PERIOD_4Y]: "4 anos",
  [PERIOD_5Y]: "5 anos",
};

export const ALERT_MILESTONES = [20, 15, 7, 3, 2, 1] as const;

export const STATUS_PENDING = "pending";
export const STATUS_ACTIVE = "active";
export const STATUS_GRACE = "grace";
export const STATUS_BLOCKED = "blocked";
export const STATUS_CANCELLED = "cancelled";
export const STATUS_REVOKED = "revoked";
export const STATUS_EXPIRED = "expired";

const API_PRODUCT_ALIASES: Record<string, string> = {
  cloud: PRODUCT_CLOUD,
  erp: PRODUCT_CLOUD,
  excellence: PRODUCT_CLOUD,
  excellence_cloud: PRODUCT_CLOUD,
  lab: PRODUCT_LAB,
  dental_lab: PRODUCT_LAB,
  vde: PRODUCT_VDE,
  vde_incorporadora: PRODUCT_VDE,
  limpeza: PRODUCT_LIMPEZA,
  cleaning: PRODUCT_LIMPEZA,
  script_limpeza: PRODUCT_LIMPEZA,
};

export function nowUtc(): Date {
  return new Date();
}

export function generateLicenseKey(): string {
  const bytes = randomBytes(LICENSE_KEY_LEN);
  let key = "";
  for (let i = 0; i < LICENSE_KEY_LEN; i++) {
    key += LICENSE_ALPHABET[bytes[i]! % LICENSE_ALPHABET.length];
  }
  return key;
}

export function generateLabSecret(): string {
  return randomBytes(32).toString("hex");
}

export function normalizeLicenseKey(key: string | null | undefined): string {
  return (key ?? "").replace(/[^A-Za-z0-9]/g, "").toUpperCase().slice(0, LICENSE_KEY_LEN);
}

export function isValidLicenseKeyFormat(key: string): boolean {
  const normalized = normalizeLicenseKey(key);
  return new RegExp(`^[A-Z0-9]{${LICENSE_KEY_LEN}}$`).test(normalized);
}

export function computeEndsAt(start: Date, periodo: string): Date {
  const days = PERIOD_DAYS[periodo] ?? PERIOD_DAYS[PERIOD_TRIAL]!;
  const end = new Date(start);
  end.setUTCDate(end.getUTCDate() + days);
  return end;
}

export function normalizeApiProduct(product: string): string {
  const raw = (product ?? "").trim().toLowerCase();
  return API_PRODUCT_ALIASES[raw] ?? raw;
}

export function productMatchesLicense(apiProduct: string, licenseProduto: string): boolean {
  const normalized = normalizeApiProduct(apiProduct);
  const legacy = licenseProduto.trim().toLowerCase();
  if (legacy === "cloud_lab") return false;
  return normalized === legacy;
}

export function daysUntil(target: Date | null | undefined, ref: Date): number {
  if (!target) return 0;
  const t = Date.UTC(target.getUTCFullYear(), target.getUTCMonth(), target.getUTCDate());
  const r = Date.UTC(ref.getUTCFullYear(), ref.getUTCMonth(), ref.getUTCDate());
  return Math.max(0, Math.floor((t - r) / 86400000));
}

export type LicenseValidity = {
  licenseStatus: string;
  licenseExpired: boolean;
  daysRemaining: number;
  validForSoftware: boolean;
};

export function computeLicenseValidity(params: {
  manualStatus: string;
  endsAt: Date | string | null | undefined;
  now?: Date;
}): LicenseValidity {
  const ref = params.now ?? nowUtc();
  const { manualStatus, endsAt } = params;

  if (manualStatus === STATUS_CANCELLED || manualStatus === STATUS_REVOKED) {
    return { licenseStatus: manualStatus, licenseExpired: true, daysRemaining: 0, validForSoftware: false };
  }

  const due = parseTs(endsAt);
  if (!due) {
    return { licenseStatus: STATUS_PENDING, licenseExpired: false, daysRemaining: 0, validForSoftware: false };
  }

  const remaining = daysUntil(due, ref);
  const expired = ref > due;
  return {
    licenseStatus: expired ? STATUS_EXPIRED : STATUS_ACTIVE,
    licenseExpired: expired,
    daysRemaining: expired ? 0 : remaining,
    validForSoftware: !expired,
  };
}

export type PaymentPhase = {
  paymentPhase: string;
  paymentStatus: string;
  daysOverdue: number;
  blockAt: string | null;
  cancelEligibleAt: string | null;
  message: string;
};

export function computePaymentPhase(params: {
  manualStatus: string;
  paymentDueAt: Date | string | null | undefined;
  blockAfterDays: number;
  cancelAfterDays: number;
  now?: Date;
}): PaymentPhase {
  const ref = params.now ?? nowUtc();
  const { manualStatus, paymentDueAt, blockAfterDays, cancelAfterDays } = params;

  if (manualStatus === STATUS_CANCELLED || manualStatus === STATUS_REVOKED) {
    return {
      paymentPhase: manualStatus,
      paymentStatus: manualStatus,
      daysOverdue: 0,
      blockAt: null,
      cancelEligibleAt: null,
      message: "Cliente cancelado ou licença revogada.",
    };
  }

  const due = parseTs(paymentDueAt);
  if (!due) {
    return {
      paymentPhase: STATUS_PENDING,
      paymentStatus: STATUS_PENDING,
      daysOverdue: 0,
      blockAt: null,
      cancelEligibleAt: null,
      message: "Aguardando definição de pagamento.",
    };
  }

  const blockAt = new Date(due);
  blockAt.setUTCDate(blockAt.getUTCDate() + blockAfterDays);
  const cancelAt = new Date(due);
  cancelAt.setUTCDate(cancelAt.getUTCDate() + cancelAfterDays);
  const daysOverdue = Math.max(0, Math.floor((ref.getTime() - due.getTime()) / 86400000));

  if (ref <= due) {
    return {
      paymentPhase: STATUS_ACTIVE,
      paymentStatus: STATUS_ACTIVE,
      daysOverdue: 0,
      blockAt: blockAt.toISOString(),
      cancelEligibleAt: cancelAt.toISOString(),
      message: "Pagamento em dia.",
    };
  }

  if (ref <= blockAt) {
    const graceLeft = blockAfterDays - daysOverdue;
    return {
      paymentPhase: STATUS_GRACE,
      paymentStatus: STATUS_GRACE,
      daysOverdue,
      blockAt: blockAt.toISOString(),
      cancelEligibleAt: cancelAt.toISOString(),
      message: `Pagamento vencido há ${daysOverdue} dia(s). Carência comercial: ${graceLeft} dia(s) restante(s).`,
    };
  }

  if (ref <= cancelAt) {
    return {
      paymentPhase: STATUS_BLOCKED,
      paymentStatus: STATUS_BLOCKED,
      daysOverdue,
      blockAt: blockAt.toISOString(),
      cancelEligibleAt: cancelAt.toISOString(),
      message: `Bloqueio comercial por inadimplência (${daysOverdue} dias após vencimento do pagamento).`,
    };
  }

  return {
    paymentPhase: "cancel_eligible",
    paymentStatus: STATUS_BLOCKED,
    daysOverdue,
    blockAt: blockAt.toISOString(),
    cancelEligibleAt: cancelAt.toISOString(),
    message: `Elegível para cancelamento comercial (${daysOverdue} dias). Limite: ${cancelAfterDays} dias após vencimento do pagamento.`,
  };
}

export function computeAlertLevel(params: {
  licenseExpired: boolean;
  daysRemaining: number;
  paymentPhase: string;
}): string {
  if (params.licenseExpired) return "expired";
  if (params.paymentPhase === STATUS_BLOCKED || params.paymentPhase === "cancel_eligible") return "critical";
  if (params.paymentPhase === STATUS_GRACE) return "warning";
  if ((ALERT_MILESTONES as readonly number[]).includes(params.daysRemaining)) return "warning";
  if (params.daysRemaining <= 3) return "critical";
  return "none";
}

export type EffectiveStatus = LicenseValidity &
  PaymentPhase & {
    status: string;
    phase: string;
    validForSoftware: boolean;
    alertLevel: string;
    remainingLabel: string;
  };

export function computeEffectiveStatus(params: {
  manualStatus: string;
  endsAt: Date | string | null | undefined;
  paymentDueAt?: Date | string | null | undefined;
  blockAfterDays?: number;
  cancelAfterDays?: number;
  now?: Date;
}): EffectiveStatus {
  const ref = params.now ?? nowUtc();
  const blockAfterDays = params.blockAfterDays ?? 30;
  const cancelAfterDays = params.cancelAfterDays ?? 45;

  const validity = computeLicenseValidity({ manualStatus: params.manualStatus, endsAt: params.endsAt, now: ref });
  const payment = computePaymentPhase({
    manualStatus: params.manualStatus,
    paymentDueAt: params.paymentDueAt ?? params.endsAt,
    blockAfterDays,
    cancelAfterDays,
    now: ref,
  });

  let validForSoftware =
    validity.validForSoftware && params.manualStatus !== STATUS_REVOKED && params.manualStatus !== STATUS_CANCELLED;
  if (payment.paymentPhase === STATUS_BLOCKED || payment.paymentPhase === "cancel_eligible") {
    validForSoftware = false;
  }

  const alertLevel = computeAlertLevel({
    licenseExpired: validity.licenseExpired,
    daysRemaining: validity.daysRemaining,
    paymentPhase: payment.paymentPhase,
  });

  let status = payment.paymentPhase;
  if (validity.licenseExpired) status = STATUS_EXPIRED;

  return {
    ...validity,
    ...payment,
    status,
    phase: status,
    daysRemaining: validity.daysRemaining,
    daysOverdue: payment.daysOverdue,
    validForSoftware,
    alertLevel,
    remainingLabel: formatRemainingCounter(validity.daysRemaining),
  };
}

export function formatRemainingCounter(days: number): string {
  if (days <= 0) return "Vencido";
  const years = Math.floor(days / 365);
  const rem = days % 365;
  const months = Math.floor(rem / 30);
  const d = rem % 30;
  const parts: string[] = [];
  if (years) parts.push(`${years} ano${years > 1 ? "s" : ""}`);
  if (months) parts.push(`${months} mês${months > 1 ? "es" : ""}`);
  if (d || parts.length === 0) parts.push(`${d} dia${d !== 1 ? "s" : ""}`);
  return parts.join(", ");
}

function parseTs(value: Date | string | null | undefined): Date | null {
  if (value == null) return null;
  if (value instanceof Date) return Number.isNaN(value.getTime()) ? null : value;
  const raw = String(value).trim().replace("T", " ").replace("Z", "");
  const parsed = new Date(raw.includes(" ") ? raw : `${raw}T00:00:00Z`);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}
